mining.go

```diff
@@ -2055,3 +2035,308 @@ func (g *BlkTmplGenerator) UpdateBlockTime(header *wire.BlockHeader) error {
+	regenChan                  chan bool
+	tg                         *BlkTmplGenerator
+	txSource                   mining.TxSource
+	lastRegen                  int64
+	subscribedClients          map[chan *BlockTemplate]struct{}
+	miningAddrs                []dcrutil.Address
+	templateMtx                sync.Mutex
+	subscriptionMtx            sync.Mutex
+	currentTemplate            *BlockTemplate
+	parentTemplate             *BlockTemplate
+	templatePool               map[[merkleRootPairSize]byte]*wire.MsgBlock
+	templatePoolMtx            sync.RWMutex
+	started                    bool
+	startedMtx                 sync.Mutex
+	permitConnectionlessMining bool
+}
+
+// newBgBlkTmplGenerator initializes a background block template generator.
+func newBgBlkTmplGenerator(tg *BlkTmplGenerator, mininingAddrs []dcrutil.Address, permitConnectionlessMining bool) *BgBlkTmplGenerator {
+	return &BgBlkTmplGenerator{
+		msgChan:                    make(chan interface{}),
+		regenChan:                  make(chan bool, 1),
+		tg:                         tg,
+		txSource:                   tg.txSource,
+		miningAddrs:                mininingAddrs,
+		subscribedClients:          make(map[chan *BlockTemplate]struct{}),
+		templatePool:               make(map[[merkleRootPairSize]byte]*wire.MsgBlock),
+		permitConnectionlessMining: permitConnectionlessMining,
+	}
+}
+
+// NotifySubscribedTemplateClients broadcasts the provided block template
+// to all registered clients. All subscribed clients are unsubscribed after
+// being sent a template update. Clients are required to resubscribe after a
+// template update.
+func (g *BgBlkTmplGenerator) NotifySubscribedTemplateClients(t *BlockTemplate) {
+	g.subscriptionMtx.Lock()
+	for c := range g.subscribedClients {
+		c <- t
+		// Delete the client after sending an update.
+		delete(g.subscribedClients, c)
+	}
+	g.subscriptionMtx.Unlock()
+}
+
+// RequestTemplateUpdate subscribes a client for a block template update. The
+// client is unsubscribed after being updated and required to resubscribe
+// for the next block template update.
+func (g *BgBlkTmplGenerator) RequestTemplateUpdate() chan *BlockTemplate {
+	updateChan := make(chan *BlockTemplate)
+	g.subscriptionMtx.Lock()
+	g.subscribedClients[updateChan] = struct{}{}
+	g.subscriptionMtx.Unlock()
+	return updateChan
+}
+
+// regenTemplate regenerates the block template. This should be run as a
+// goroutine.
+func (g *BgBlkTmplGenerator) regenTemplate() {
+	// NOTE: If the blockchain package is updated to provide more
+	// fine grained locking, it will be necessary to add a call to
+	// WaitForChain here.
+
+	// Regenerate block templates on mainnet only when the chain is synced.
+	if !g.tg.chain.IsCurrent() && !g.permitConnectionlessMining {
+		return
+	}
+
+	// Pick a mining address at random.
+	rand.Seed(time.Now().UnixNano())
+	payToAddr := g.miningAddrs[rand.Intn(len(g.miningAddrs))]
+
+	// Regenerate the block template.
+	g.templateMtx.Lock()
```

Comment:
This needs to be after the block template is generated or it will result in a deadlock since `NewBlockTempalate` still relies on sending messages to `blockmanager` and the notification handler in the same goroutine also tries to take the same lock.

```Go
	// Regenerate the block template.
	template, err := g.tg.NewBlockTemplate(payToAddr)
	if err != nil {
		minrLog.Debugf("block template generation failed: %v", err)
		return
	}

	g.templateMtx.Lock()
	g.currentTemplate = template
	...
```